### PR TITLE
Hometask 8 - Http Client Module, Interceptor and Backend integration

### PR DESCRIPTION
**Pull request by Jonatan-Alvarez-EPAM**
Http header request with authorization token:

token

### CODE REVIEW

File: src/app/app.module.ts [hunk 1]
```diff
@@ -15,9 +17,12 @@ import { AppComponentsRoutingModule } from './app-components/app-components-rout
     AppLayoutModule,
     AppComponentsRoutingModule,
     AppRoutingModule,
+    HttpClientModule,
   ],
   providers: [
     { provide: 'Storage', useValue: localStorage },
+    { provide: 'BASE_URL', useValue: 'http://localhost:3004' },
```
Comment (aledpardo) [inline src/app/app.module.ts]: This is a great way to provide the URL! Very well done!

File: src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.html [hunk 2]
```diff
@@ -1,4 +1,4 @@
 <div class="rectangle">
     <span><a routerLink="/courses">Courses</a></span>
-    <span *ngIf="courseId">&nbsp;/&nbsp;{{courseTitle}}</span>
+    <span *ngIf="courseId">&nbsp;/&nbsp;{{courseName}}</span>
```
Comment (aledpardo) [inline src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.html]: We may use CSS paddings to add spacing between HTML elements. For example:

```html
<span>
    <a routerLink="/courses">Courses</a>
</span>

<ng-container *ngIf="courseId">
    <span class='breadcrumb-separator'>/</span>
    <span>{{courseName}}</span>
</ng-container>
```

```css
.breadcrumb-separator{
    padding: 0 .5rem; // or 0 2px;
}
```

File: src/app/app-components/courses/add-course/add-course.component.ts [hunk 3]
```diff
@@ -42,18 +51,34 @@ export class AddCourseComponent implements OnInit {
...
-      this.coursesService.create(this.courseInfo);
+      // ToDo: Implement create course in BE
+      this.httpClient.post(`${this.BASE_URL}/courses`, {
+        id: this.courseInfo.id,
+        name: this.courseInfo.name,
+        date: this.courseInfo.date,
+        length: this.courseInfo.length,
+        description: this.courseInfo.description,
+        isTopRated: this.courseInfo.isTopRated,
+      }).subscribe();
```
Comment (aledpardo) [inline src/app/app-components/courses/add-course/add-course.component.ts]: We can use the many `subscribe` callbacks to handle the result of the HTTP call. For example:

```typescript
...
.subscribe(
    (response) => { /* handle success */},
    (error) => { /* handle the error */ }
);
```

File: src/app/app-components/courses/courses-list/courses-list.component.ts [hunk 4]
```diff
@@ -12,30 +14,55 @@ import { CoursesService } from '@app/app-services';
...
+        return this.loadByPage(this.previousPage(), this.nextPage())
+          .pipe(map((courses: Course[]) => new OrderByPipe().transform(courses)));
+      }));
     }
   }
 
   onLoadMore() {
-    console.log('[FAKE] Load more...');
+    const result$ = this.loadByPage(this.previousPage(), this.nextPage());
+    this.courses$ = zip(this.courses$, result$)
```
Comment (aledpardo) [inline src/app/app-components/courses/courses-list/courses-list.component.ts]: Well done using the `zip` operator!! 🥇

File: src/app/app-services/AuthInterceptor.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import { HttpInterceptor, HttpHandler, HttpRequest, HttpEvent } from "@angular/common/http";
+import { Observable } from 'rxjs';
+import { AuthorizationService } from './authorization.service';
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+    constructor(
+        private readonly authService: AuthorizationService,
+        @Inject('Storage') private readonly localStorage: Storage) { }
+    intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
```
Comment (aledpardo) [inline src/app/app-services/AuthInterceptor.ts]: Just as a tip, here's how I usually format the code when a method has many arguments:

```typescript
constructor(
    private readonly authService: AuthorizationService,
    @Inject('Storage') private readonly localStorage: Storage
)  { }

intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    // code
}
```

File: src/app/app-services/authorization.service.ts [hunk 6]
```diff
@@ -1,45 +1,54 @@
...
 
-  private readonly KEY = 'CURRENT_USER';
+  readonly USER = 'CURRENT_USER';
+  readonly TOKEN = 'TOKEN';
+  token?: string;
 
-  constructor(@Inject('Storage') private readonly localStorage: Storage) {
+  constructor(@Inject('Storage') private readonly localStorage: Storage,
+    @Inject('BASE_URL') private readonly BASE_URL: string,
+    private readonly httpClient: HttpClient) {
```
Comment (aledpardo) [inline src/app/app-services/authorization.service.ts]: One more tip to format methods with many arguments:

```typescript
constructor(
    @Inject('Storage') private readonly localStorage: Storage,
    @Inject('BASE_URL') private readonly BASE_URL: string,
    private readonly httpClient: HttpClient
) {}
```

File: src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.html (hunk 2 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.html]: Done.

File: src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.ts [hunk 7]
```diff
@@ -11,16 +13,21 @@ import { CoursesService } from '@app/app-services';
...
+    private readonly httpClient: HttpClient,
+    @Inject('BASE_URL') private readonly BASE_URL: string) {
   }
 
   ngOnInit() {
     this.courseId = this.route.snapshot.paramMap.get('id');
     if (this.courseId) {
-      const course = this.coursesService.get(this.courseId);
-      this.courseTitle = course.title;
+      this.httpClient.get<Course>(`${this.BASE_URL}/courses`, {
```
[the remark this answers is not in the document]
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.ts]: Done.

File: src/app/app-components/courses/add-course/add-course.component.ts (hunk 3 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-components/courses/add-course/add-course.component.ts]: Done.

File: src/app/app-services/AuthInterceptor.ts (hunk 5 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-services/AuthInterceptor.ts]: Thanks for the tip, I was not sure about how to format all this params.

File: src/app/app-services/authorization.service.ts (hunk 6 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-services/authorization.service.ts]: Ditto

**Comment by Jonatan-Alvarez-EPAM** [comment]
Hi Ale,

Please feel free to check my second commit on this same branch. I addressed all your comments and fixed some minor issues I found.

Thanks,
Jon

File: src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.scss [hunk 8]
```diff
@@ -6,4 +6,9 @@
   background-color: var(--white);
   align-items: center;
   padding: 0 25px;
+}
+
+.course-name::before {
+    content: '/';
```
Comment (aledpardo) [inline src/app/app-components/breadcrumbs-navbar/breadcrumbs-navbar.component.scss]: Nice solution!!

File: src/app/app-services/courses.service.ts [hunk 9]
```diff
@@ -1,83 +1,49 @@
...
+    return this.httpClient.get<Course[]>(`${this.BASE_URL}/courses`, { params });
   }
 
-  get(idCourse: string): Course | undefined {
-    const index = this.find(idCourse);
-    if (Number.isInteger(index)) {
-      return this.courses[index];
-    }
+  create(courseInfo: Course): Observable<Course> {
+    return this.httpClient.post<Course>(`${this.BASE_URL}/courses`, {
```
Comment (aledpardo) [inline src/app/app-services/courses.service.ts]: As the data types of the `create` method and the API are the same, we may simply pass to `post` the `courseInfo` object instead.

```typescript
return this.httpClient.post<Course>(`${this.BASE_URL}/courses`, courseInfo);
```

File: src/app/app-services/courses.service.ts [hunk 10]
```diff
@@ -1,83 +1,49 @@
...
+  get(idCourse: string): Observable<Course> {
+    return this.httpClient.get<Course>(`${this.BASE_URL}/courses`, { params: { id: idCourse } });
   }
 
-  delete(idCourse: string) {
-    const index = this.find(idCourse);
-    if (Number.isInteger(index)) {
-      this.courses.splice(index, 1);
-    }
+  update(courseInfo: Course): Observable<Course> {
```
Comment (aledpardo) [inline src/app/app-services/courses.service.ts]: What I said in the `create` method applies here as well.

File: src/app/app-services/courses.service.ts (hunk 9 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-services/courses.service.ts]: Done. (Thanks for catching this :smiley: )

File: src/app/app-services/courses.service.ts (hunk 10 above)
Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-services/courses.service.ts]: Ditto

Comment (Jonatan-Alvarez-EPAM) [reply src/app/app-services/courses.service.ts]: Ditto

**Pull request merged**